packages/grid/_modules_/grid/hooks/features/infiniteLoader/useGridInfiniteLoader.ts

```diff
@@ -2,50 +2,103 @@ import * as React from 'react';
+} from '../../../constants/eventsConstants';
 import { gridContainerSizesSelector } from '../../root/gridContainerSizesSelector';
 import { useGridApiEventHandler, useGridApiOptionHandler } from '../../root/useGridApiEventHandler';
 import { GridRowScrollEndParams } from '../../../models/params/gridRowScrollEndParams';
 import { visibleGridColumnsSelector } from '../columns/gridColumnsSelector';
 import { GridComponentProps } from '../../../GridComponentProps';
+import { renderStateSelector } from '../virtualization/renderingStateSelector';
+import { GridViewportRowsChangeParams } from '../../../models/params/gridViewportRowsChangeParams';
+import { GridScrollParams } from '../../../models/params/gridScrollParams';
 
 export const useGridInfiniteLoader = (
   apiRef: GridApiRef,
-  props: Pick<GridComponentProps, 'onRowsScrollEnd'>,
+  props: Pick<GridComponentProps, 'onRowsScrollEnd' | 'onViewportRowsChange'>,
 ): void => {
   const options = useGridSelector(apiRef, optionsSelector);
   const containerSizes = useGridSelector(apiRef, gridContainerSizesSelector);
   const visibleColumns = useGridSelector(apiRef, visibleGridColumnsSelector);
   const isInScrollBottomArea = React.useRef<boolean>(false);
+  const renderState = useGridSelector(apiRef, renderStateSelector);
+  const previousRenderContext = React.useRef<null | {
+    firstRowIndex: number;
+    lastRowIndex: number;
+  }>(null);
+
+  const handleRowsScrollEnd = React.useCallback(
+    (scrollPosition: GridScrollParams) => {
+      if (!containerSizes) {
+        return;
+      }
+
+      const scrollPositionBottom =
+        scrollPosition.top + containerSizes.windowSizes.height + options.scrollEndThreshold;
+
+      if (scrollPositionBottom < containerSizes.dataContainerSizes.height) {
+        isInScrollBottomArea.current = false;
+      }
+
+      if (
+        scrollPositionBottom >= containerSizes.dataContainerSizes.height &&
+        !isInScrollBottomArea.current
+      ) {
+        const rowScrollEndParam: GridRowScrollEndParams = {
+          api: apiRef,
+          visibleColumns,
+          viewportPageSize: containerSizes.viewportPageSize,
+          virtualRowsCount: containerSizes.virtualRowsCount,
+        };
+        apiRef.current.publishEvent(GRID_ROWS_SCROLL_END, rowScrollEndParam);
+        isInScrollBottomArea.current = true;
+      }
+    },
+    [apiRef, options, visibleColumns, containerSizes],
+  );
 
   const handleGridScroll = React.useCallback(() => {
-    if (!containerSizes) {
+    if (!props.onRowsScrollEnd && !props.onViewportRowsChange) {
       return;
     }
 
     const scrollPosition = apiRef.current.getScrollPosition();
-    const scrollPositionBottom =
-      scrollPosition.top + containerSizes.windowSizes.height + options.scrollEndThreshold;
 
-    if (scrollPositionBottom < containerSizes.dataContainerSizes.height) {
-      isInScrollBottomArea.current = false;
+    if (props.onRowsScrollEnd) {
+      handleRowsScrollEnd(scrollPosition);
+    }
```

Comment:
Same feedback as in https://github.com/mui-org/material-ui-x/pull/2038/files#r683870957

```suggestion
    handleRowsScrollEnd(scrollPosition);
```